sandbox/rocky/tf/optimizers/conjugate_gradient_optimizer.py

```diff
@@ -7,47 +7,51 @@
 import numpy as np
 import tensorflow as tf
 from sandbox.rocky.tf.misc import tensor_utils
+from sandbox.rocky.tf.misc.tensor_utils import sys_op_scope
 from rllab.misc.ext import sliced_fun
 
 
 class PerlmutterHvp(object):
-    def __init__(self, num_slices=1):
+    def __init__(self, num_slices=1, name="PerlmutterHvp"):
         self.target = None
         self.reg_coeff = None
         self.opt_fun = None
         self._num_slices = num_slices
-
-    def update_opt(self, f, target, inputs, reg_coeff):
-        self.target = target
-        self.reg_coeff = reg_coeff
-        params = target.get_params(trainable=True)
-
-        constraint_grads = tf.gradients(f, xs=params)
-        for idx, (grad, param) in enumerate(zip(constraint_grads, params)):
-            if grad is None:
-                constraint_grads[idx] = tf.zeros_like(param)
-
-        xs = tuple([tensor_utils.new_tensor_like(p.name.split(":")[0], p) for p in params])
-
-        def Hx_plain():
-            Hx_plain_splits = tf.gradients(
-                tf.reduce_sum(
-                    tf.stack([tf.reduce_sum(g * x) for g, x in zip(constraint_grads, xs)])
+        self._name = name
+
+    def update_opt(self, f, target, inputs, reg_coeff, name="update_opt"):
+        with sys_op_scope(self._name, name):
+            self.target = target
+            self.reg_coeff = reg_coeff
+            params = target.get_params(trainable=True)
+
+            constraint_grads = tf.gradients(f, xs=params, name="constraint_gradients")
+            for idx, (grad, param) in enumerate(zip(constraint_grads, params)):
+                if grad is None:
+                    constraint_grads[idx] = tf.zeros_like(param)
+
+            xs = tuple([tensor_utils.new_tensor_like(p.name.split(":")[0], p) for p in params])
+
+            def Hx_plain():
+                Hx_plain_splits = tf.gradients(
```

Comment:
Although the tensor for Hx_plain_gradients is defined within the class scope of PerlmutterHvp, function Hx_plain() does not get call within the class, and therefore the variable scope of the primitive does not apply to Hx_plain_gradients

Reply:
Hmm this just seems to me like a special case with clear correct behavior. We can use the closure property to capture the correct enclosing scope during function definition. Then it doesn't matter whether it's called in a lambda or not.
```python
def Hx_plain():
    with sys_op_scope(self._name, "Hx_plain")
        Hx_plain_splits = tf.gradients( #...
```
And then even when it's invoked via lambda, self._name will have been captured during the definition and it will appear inside scope PerlmutterHvp.

Reply:
Oh, I guess context managers don't cross function boundaries. That makes sense.

what if we did
```python
def Hx_plain():
    with sys_op_scope(self._name, name):
        with sys_op_scope(name, "Hx_plain"):
            Hx_plain_splits = tf.gradients( #...
```

I think that would make the expected output.